[AN-Issue-1346] Implemented custom handling of automation transaction

aregng opened the pull request:
- Added separate flow for automation transaction handling. It differes from user transaction flow only by custom validation step of inner-payload of automation transaction

- Added validation logic for the automation transaction inner-payload

- Introuced specific StatusCode for failures during automation transaction inner-payload validation. They are qualified as verification status type and will lead to have transaction kept and the account will be charged.

Commit:
[AN-Issue-1346] Implemented custom handling of automation transactions

- Added separate flow for automation transaction handling. It differes
from user transaction flow only by custom validation step of
inner-payload of automation transaction

- Added validation logic for the automation transaction inner-payload

- Introuced specific StatusCode for failures during automation
transaction inner-payload validation. They are qualified as verification
status type and will lead to have transaction kept and the account will
be charged.

aregng commented on types/src/transaction/automation.rs:
```diff
@@ -0,0 +1,58 @@
+
+/// Represents options for Automation transaction payload.
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub enum AutomationTransactionPayload {
```
Taking into account that we are doing already custom handling for the user transactions which aim automation task registration here we can have another option as well, instead of EntryFunction with call to  `automation_registry::register` we can have struct enclosing all arguments that are required to pass to the register function.
```
AutomationTransactionPayload {
        automated_txn_payload: EntryFunction,
        expiry_time: u64,
        max_gas_amount: u64,
        gas_price_cap: u64,
}
```
Then during transaction execution we can easily verify `automated_txn_payload` and then manually construct  a new entry-function to call `automation_registry::register` function with the above properties as arguments to the function.

With this approach this type of validity checked here https://github.com/Entropy-Foundation/aptos-core/pull/134/files#diff-fd22752e67147bf54afdcfa60f8074d0d6558f1610b34a940e6591d7bbdaec45R902 can be avoided.

Dr. @sjoshisupra what are your thoughts related to this.

sjoshisupra replied:
@aregng , downside is that if tomorrow we are to change the signature of the function (of course we have to rename it, since we can't republish the same function with different signature), then the change at Rust layer would be much more. I would prefer just checking the target function is an EntryFunction for registration. In case of a change, only change of name would be required in Rust layer.

aregng commented on aptos-move/aptos-vm/src/transaction_metadata.rs:
```diff
@@ -34,6 +35,7 @@ pub struct TransactionMetadata {
     pub is_keyless: bool,
     pub entry_function_payload: Option<EntryFunction>,
     pub multisig_payload: Option<Multisig>,
+    pub automation_payload: Option<AutomationTransactionPayload>,
```
if we do not see a need to have access to `AutomationTransactionPayload` from move-context, then this can be simply substituted with boolean flag indicating the origin of the entry-function call.

sjoshisupra replied:
@aregng , I was wondering if instead of a `bool`, we can have an `enum`  that indicates the originating tx type. That way tomorrow if it is a tx of a different kind and we need to identify from `SafeNativeContext`, we can do so without adding one more field here.

Commit:
Updated copyright info

sjoshisupra commented on api/src/transactions.rs:
```diff
@@ -1,6 +1,8 @@
 // Copyright © Aptos Foundation
 // Parts of the project are originally copyright © Meta Platforms, Inc.
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright ©  2024 Supra.
// Copyright © Aptos Foundation
// Parts of the project are originally copyright © Meta Platforms, Inc.
// SPDX-License-Identifier: Apache-2.0
//

```

sjoshisupra commented on aptos-move/aptos-debugger/src/aptos_debugger.rs:
```diff
@@ -1,9 +1,12 @@
 // Copyright © Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright  © 2024 Supra.
// Copyright © Aptos Foundation
// SPDX-License-Identifier: Apache-2.0
//

```

sjoshisupra commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -878,6 +879,87 @@ impl AptosVM {
+        }
+
+        gas_meter.charge_intrinsic_gas_for_transaction(txn_data.transaction_size())?;
+        if txn_data.is_keyless() {
+            gas_meter.charge_keyless()?;
+        }
```
Has this been copied from some existing function? As of now we don't support keyless logins.

aregng replied:
yes, I kept the same flow as for SignedTranansaction which is not considered non-multisig. So if we do not support it now, I just kept it for future.

sjoshisupra commented on aptos-move/e2e-tests/src/executor.rs:
```diff
@@ -1,6 +1,8 @@
 // Copyright © Aptos Foundation
 // Parts of the project are originally copyright © Meta Platforms, Inc.
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright © 2024 Supra.
// Copyright © Aptos Foundation
// Parts of the project are originally copyright © Meta Platforms, Inc.
// SPDX-License-Identifier: Apache-2.0
//

```

sjoshisupra commented on aptos-move/framework/src/natives/transaction_context.rs:
```diff
@@ -395,6 +395,34 @@ fn native_multisig_payload_internal(
+    context: &mut SafeNativeContext,
+    mut _ty_args: Vec<Type>,
+    _args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
```
Not sure what exactly we are returning to Move layer.

aregng replied:
will fix the implementation and add a comment.

sjoshisupra commented on consensus/src/transaction_shuffler/fairness/conflict_key/entry_fun.rs:
```diff
@@ -1,7 +1,10 @@
 // Copyright © Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright © 2024 Supra.
// Copyright © Aptos Foundation
// SPDX-License-Identifier: Apache-2.0
//

```

sjoshisupra commented on consensus/src/transaction_shuffler/fairness/conflict_key/entry_fun_module.rs:
```diff
@@ -1,7 +1,10 @@
 // Copyright © Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright © 2024 Supra.
// Copyright © Aptos Foundation
// SPDX-License-Identifier: Apache-2.0
//

```

sjoshisupra commented on third_party/move/move-core/types/src/vm_status.rs:
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) The Diem Core Contributors
 // Copyright (c) The Move Contributors
 // SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright (c) 2024 Supra.
// Copyright (c) The Diem Core Contributors
// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0
//

```

Commit:
Added language-e2e-testsuite to ci-flow with fixed tests

Commit:
Addressed review comments and added tests

- Introduced a new type PayloadTypeReference which will describe
  user-transaction payload type and reference payload if required.

- Updated TransactionMetadata and UserTransctionContext to utilize
  PayloadTypeReference enabling easy extension of any new payload type
  when needed

- Updated
  transaction_arg_validation::validate_combine_signer_and_txn_args to
  accept GasMeter as input argument.

- Updated automation inner payload validation to charge gas for it

- Added initial set of e2e tests to check Automation transaction flow.

Commit:
Fixed test failures after rebase

Commit:
Fixed copyright comment

Commit:
Introduce means to create and process automation transaction specifying the required arguments for registration

sjoshisupra commented on api/types/src/convert.rs:
```diff
@@ -278,65 +281,20 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
-                            };
-
+                            let entry_function_payload =
+                                self.try_to_entry_function_payload(entry_function)?;
```
Did you mean `try_into_entry_function_payload`? If this is a type coercion operation, I believe the convetion is to use `into`.

sjoshisupra commented on api/types/src/convert.rs:
```diff
@@ -656,40 +629,8 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
-                        .collect::<Result<_>>()?,
-                    args,
-                ))
+                let entry_function = self.try_to_aptos_core_entry_function(entry_func_payload)?;
```
Same suggestion, if this is for type coercion, perhaps we should use `into`. If this is a function define by us, perhaps we may use `supra_core` instead?

sjoshisupra commented on api/types/src/convert.rs:
```diff
@@ -720,44 +661,11 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
-                                .map(bcs::to_bytes)
-                                .collect::<Result<_, bcs::Error>>()?;
+                            let entry_function =
+                                self.try_to_aptos_core_entry_function(entry_function)?;
```
Same suggestion as earlier.

sjoshisupra commented on api/types/src/convert.rs:
```diff
@@ -1082,6 +1007,72 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
         Ok(id.to_string())
     }
+
+    fn try_to_entry_function_payload(&self, fun: EntryFunction) -> Result<EntryFunctionPayload> {
```
```suggestion
    fn try_into_entry_function_payload(&self, fun: EntryFunction) -> Result<EntryFunctionPayload> {
```

sjoshisupra commented on api/types/src/convert.rs:
```diff
@@ -1082,6 +1007,72 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
+        })
+    }
+
+    fn try_to_aptos_core_entry_function(
```
If the function name `try_into_entry_function` doesn't already exist, perhaps it may be used.

```suggestion
    fn try_into_entry_function(
```

aregng replied:
They are 2 different functions one converts `api::EntryFucntion->core::EntryFucntion` and the other does the reverse conversion.

sjoshisupra commented on api/types/src/transaction.rs:
```diff
@@ -1061,6 +1063,24 @@ impl VerifyInput for MultisigPayload {
 
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct AutomationTransactionPayload {
+    pub inner_payload: EntryFunctionPayload,
```
At move layer this would be `Vec<u8>`, do we have appropriate conversion when making a call to `register` method in Move layer?

aregng replied:
yes it is correctly converted.
https://github.com/Entropy-Foundation/aptos-core/blob/87dab4b4dfa179d188868881eb229ce7be6a10c1/types/src/transaction/automation.rs#L59

sjoshisupra commented on aptos-move/aptos-debugger/src/aptos_debugger.rs:
```diff
@@ -99,6 +100,12 @@ impl AptosDebugger {
                 let gas_profiler = match txn.payload() {
                     TransactionPayload::Script(_) => GasProfiler::new_script(gas_meter),
+                    TransactionPayload::Automation(auto_payload) => GasProfiler::new_function(
+                        gas_meter,
```
Is this `Unmertered` type? Or does this `gas_meter` actually meter positive gas?

aregng replied:
Yes it will meter positive gas utilizing production gas-meter
- https://github.com/Entropy-Foundation/aptos-core/blob/87dab4b4dfa179d188868881eb229ce7be6a10c1/aptos-move/aptos-vm/src/aptos_vm.rs#L1949
- https://github.com/Entropy-Foundation/aptos-core/blob/87dab4b4dfa179d188868881eb229ce7be6a10c1/aptos-move/aptos-vm/src/gas.rs#L107

sjoshisupra commented on aptos-move/aptos-vm/src/verifier/view_function.rs:
```diff
@@ -56,6 +58,7 @@ pub(crate) fn validate_view_function(
     let allowed_structs = get_allowed_structs(struct_constructors_feature);
     let args = transaction_arg_validation::construct_args(
         session,
+        &mut UnmeteredGasMeter,
```
View functions are typically not charged any gas on their own (unless called from some entry function). I am hoping the metering would not change that.

aregng replied:
previously construct_args where not directly accepting gas-meter as argument, but it was creating `UnmeteredGasMeter` internally to do the construction. But as long as we wanted to charge for `AutomationTransaction` embedded payload validation, I have updated the functions to accept gas-meter as input. So here specifying `UnbeteredGasMeter` does not affect affect original view-function economy and no charges will be applied.

sjoshisupra commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -877,6 +880,109 @@ impl AptosVM {
+                    )
+                })?
+            },
+            AutomationTransactionPayload::EntryFunctionArguments(args) => {
+                session.execute(|session| {
+                    self.validate_automation_txn_inner_payload_v2(
+                        session,
+                        gas_meter,
+                        txn_data.senders(),
+                        args.inner_payload(),
+                    )
+                })?;
```
Ok, my understanding is that we are either going to keep this case, or the previous one (where we have to make an assumption that payload is always the first argument in both Rust and Move layer). I agree that such an assumption would make the design fragile, so it's best to contruct it from the arguments.

sjoshisupra commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2257,14 +2375,14 @@ impl AptosVM {
-            },
+            TransactionPayload::Script(_)
+            | TransactionPayload::EntryFunction(_)
+            | TransactionPayload::Automation(_) => transaction_validation::run_script_prologue(
```
I believe we need to customize here. We have to call a different prologue, the one which does not check for `sequence_number` of `Account`.

aregng replied:
Automation is the payload of the transaction that aims to register the automation task, so i named it/refer to it  `AutomationTransaction`. So this type of transactions is considered normal user transaction with single exception which is that inner-payload/automation-task-payload should be validated. so senders sequence number will be increased in this case.
`AutoamtedTransaction` is the transaction that will be constructed based on the registered tasks, and for this transactions prologue and epologue will be different, and sequence number check and increment will be omitted: 
The draft PR for `AutomatedTranaction` processing flow can be found here:  https://github.com/Entropy-Foundation/aptos-core/pull/136

sjoshisupra replied:
@aregng , Ah I see. I wonder if in that case we should rename `AutomationTransaction` to `AutomationRegistration`. The names `AutomationTransaction` and `AutomatedTransaction` are so similar that it may be hard for a code reader to read/comprehend the code easily. What do you think?

aregng replied:
Sounds good.

sjoshisupra commented on aptos-move/e2e-tests/src/executor.rs:
```diff
@@ -684,6 +685,12 @@ impl FakeExecutor {
             |gas_meter| {
                 let gas_profiler = match txn.payload() {
                     TransactionPayload::Script(_) => GasProfiler::new_script(gas_meter),
+                    TransactionPayload::Automation(auto_payload) => GasProfiler::new_function(
+                        gas_meter,
+                        auto_payload.module_id().clone(),
+                        auto_payload.function().to_owned(),
+                        auto_payload.ty_args().to_vec(),
+                    ),
```
Is this here for gas estimation?

aregng replied:
I do not think so. `GasProfiler` simply records gas-cost for predefined set of operations along with gas-cost calculation based on specified meter, and mainly utilized in `aptos-debuger` and here in `FakeExecutor` for e2e tests

sjoshisupra commented on aptos-move/e2e-testsuite/src/tests/automation_transaction.rs:
```diff
@@ -0,0 +1,343 @@
+            .into_entry_function();
+
+    let view_output = test_context.execute_view_function(
+        str::parse("0x1::timestamp::now_seconds").unwrap(),
```
It is best to define constant literals and use it here instead of directly using the strings.

sjoshisupra commented on aptos-move/e2e-testsuite/src/tests/automation_transaction.rs:
```diff
@@ -0,0 +1,343 @@
+
+    // Check automation registry state.
+    let view_output = test_context.execute_view_function(
+        str::parse("0x1::automation_registry::get_next_task_index").unwrap(),
```
Best to define constant literals for these at one place and use those constant literals instead of strings.

sjoshisupra commented on aptos-move/e2e-testsuite/src/tests/automation_transaction.rs:
```diff
@@ -0,0 +1,343 @@
+
+    // Check automation registry state.
+    let view_output = test_context.execute_view_function(
+        str::parse("0x1::automation_registry::get_next_task_index").unwrap(),
```
Same suggestion.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -80,7 +83,7 @@ module supra_framework::automation_registry {
 
     // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
-    fun initialize(supra_framework: &signer) {
+    public(friend) fun initialize(supra_framework: &signer) {
```
Does not necessarily have to be `(friend)` as long as we check that `signer` is indeed `supra_framework`, that way we can call this from governance script.

sjoshisupra commented on consensus/src/transaction_shuffler/fairness/conflict_key/entry_fun.rs:
```diff
@@ -14,21 +16,29 @@ pub enum EntryFunKey {
+    }
+}
+
 impl ConflictKey<SignedTransaction> for EntryFunKey {
```
I am unsure of the role of `ConflictKey` here for consensus. We can discuss.

aregng replied:
my understanding it that it is used by aptos-consensus to order/select the transactions while constructing a block to avoid conflicts as much as possible.
Here is the comment for it in code-base
```
/// `ConflictKey::extract_from(txn)` extracts a key from a transaction. For example,
/// `TxnSenderKey::extract_from(txn)` returns the transaction sender's address. The key is used by
/// the shuffler to determine whether two close by transactions conflict with each other.
///
/// `ConflictKey::conflict_exempt(&key)` returns if this specific key is exempt from conflict.
/// For example, we can exempt transaction sender 0x1, so that consecutive transactions sent by
/// 0x1 are not seen as a conflict by the shuffler.
```
This feature is used to reorder the existing transactions avoiding conflicts as much as possible and then build the block from the ordered set.
This is aptos internal logic and in smr-moonshot we are not using it.

sjoshisupra commented on third_party/move/move-core/types/src/vm_status.rs:
```diff
@@ -728,6 +729,11 @@ pub enum StatusCode {
     RESERVED_VERIFICATION_ERROR_3 = 1129,
     RESERVED_VERIFICATION_ERROR_4 = 1130,
     RESERVED_VERIFICATION_ERROR_5 = 1131,
+    // Verification errors related to automation transaction
+    // Automation Transaction payload validation failed.
+    INVALID_AUTOMATION_PAYLOAD = 1132,
+    INVALID_AUTOMATION_PAYLOAD_ARGUMENTS = 1133,
+    INVALID_AUTOMATION_INNER_PAYLOAD = 1134,
```
The lines above mentions `RESERVED_VERIFICATION_ERROR_X`, does it mean that this error codes are also referenced/provisioned somewhere else? In which case these new codes would have to be integrated there as well.

aregng replied:
I have checked they are not referenced in any other place.

sjoshisupra commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -1,3 +1,6 @@
+// Copyright (c) Aptos Foundation
```
Switch copyright line order to make in reverse chronological order of contribution.

sjoshisupra reviewed (changes requested):
Has the changes to CLI been added or planned to ensure that CLI/client can make this new transaction type?

aregng commented:
> Has the changes to CLI been added or planned to ensure that CLI/client can make this new transaction type?

That should be done in scope of smr-moonshot a new CLI command should be added to create automtion-transaction.

Commit:
Addressed review comments

Commit:
Renamed 'automation transaction' to 'automation registration'

Commit:
Updated automation::registry to be non-entry and made direct call from native layer

Commit:
Cosmetic chagnes

Commit:
Copyright fix

sjoshisupra commented on api/types/src/transaction.rs:
```diff
@@ -328,7 +328,7 @@ impl
         ),
     ) -> Self {
-        // If at some point Aptos will utilize Automation transactions then here we will use as
+        // If at some point Aptos will utilize Automated transactions then here we will use as
```
```suggestion
        // If at some point Supra will utilize Automated transactions then here we will use as
```

sjoshisupra commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -1,6 +1,3 @@
-// Copyright (c) Aptos Foundation
-// SPDX-License-Identifier: Apache-2.0
-
 // Copyright (c) 2024 Supra.
```
```suggestion
// Copyright (c) 2024 Supra.
// SPDX-License-Identifier : Apache-2.0
```

Commit:
Temporarly fixed test failure

Commit:
Split and reordered test runs in ci

Commit:
Increased nextest timeout from 2100 to 3000
